### ISSUE TITLE
Hybrid flows:  CLI plugin - deploy

### ISSUE THREAD

**Issue by kfidelak94** [forcedotcom/devops-center-roadmap#20]
New CLI plugin with initial support for the following command:

`sf deploy pipeline`

Allows deployment to be performed via the CLI such that DevOps Center can properly reflect the associated state of the pipeline

**Comment by RupertBarrow** [comment]
In Github, if we merge a pull request manually, can DevOps Center automatically recover all the stuff it needs to proceed, or should we consider triggering a GH action to call "sf deploy pipeline" to help DevOps Center resync ?

Or, on the contrary, will "sf deploy pipeline" be the command to initiate the whole operation, just replacing a manual click in DevOps Center ?

**Comment by kfidelak94** [comment]
@RupertBarrow Today, if you merge a pull request manually in GH, then when you come back to DevOps Center it will recognize that the PR was merged and guide you to complete the promotion in order to deploy the source from the already-merged branch to the corresponding stage org.  With this new CLI plugin that we'll be introducing, you'll be able to "complete the promotion" and thereby deploy the changes from the already-merged branch to the target org using the new "sf deploy pipeline" command.    Either way, that "deploy" or "complete promotion" step must be initiated by the user to perform the deployment.  First question to you:  would you prefer/expect that after the PR is merged by you in GH, that DevOps Center automatically performs the deploy? 

In this first release of the plugin, the "sf deploy pipeline" command will run on an already-merged branch. In the future we may expand the capability of the CLI command to be able to do the full "promotion", ie both the merge and the deploy. Second question to you:  do you see value in an "orchestrated" CLI command that does both the merge and deploy, to mirror the "promote" action in the GUI, or do you prefer to just do the merge yourself as one step, and then execute the deploy (or per the first question, have DevOps Center automatically execute the deploy)?

**Comment by vorno-crole** [comment]
Hi folks,
Quick question (or request?) on this feature...
If I'm running the merge and deploy externally with my own logic, can I just "tell" DOC that X story was deployed and just update it's status (ie: don't rerun the deploy!)
While it's nice for CLI to request DOC to run a deployment, for specific use-cases, it would be advantageous to take ownership of this process, while leaving DOC to own the rest (story-tracking, allowing builders to capture metadata, etc.)

**Comment by RupertBarrow** [comment]
@kfidelak94 Admins I know who are using or interested in DevOps Center want to simplify their work, i.e. : 
1) avoid having to go to GitHub if it is not necessary
2) avoid having to use the CLI
3) avoid having to use VS Code

So in answer to your first question, if for some reason the PR has been merged and closed in Github, it's probably a good idea to automate the deployment of the WI in DevOps Center. The only example where it might be useful to *not* do this is a situation where a PR has been merged but (for some reason) could not be closed. If this error condition is possible and if Github allows one to add another new commit to the PR, they we should keep the WI non-promoted.

To answer your second question, I think we should ask who (what person) will be using this command.
Admins work on their WIs and mark them "ready for promotion". They also create the associated PR/review request.
Developers will create a WI and work in VS Code to deliver code on that branch. They will create the associated PR.
Project Tech Leaders/Integration Managers will merge PRs and promote change reviews in DOC; they will asks devs for reviews in GH; I don't see them reviewing metadata delivered by admins.

So merging and promoting PRs with code will happen (for now) in Github => it would be nice for those to move on automatically in DOC. PRs with admin metadata will be handled in DOC, manually.

So I don't see who (what persona) is asking for or will be using the "sf deploy pipeline" command ?

**Comment by AzaruddinKhan** [comment]
Hi @kfidelak94 ,

Does DevOps Center support external repository hosted on Azure DevOps or Jenkins ?

Can we integrate with DevOps Center instead of GITHub.


/A

**Comment by msouther-lisc** [comment]
Trying to guess the impetus/value-add behind this CLI feature, as @RupertBarrow was earlier. 

My best guess: a GitHub Action (using a container image with the Salesforce CLI installed) can be set up to call `sf deploy pipeline` when a lead developer approves and merges a PR directly in GitHub. This avoids the need to return to DevOps Center.

**Comment by RupertBarrow** [comment]
Interesting, @msouther-lisc.

**Comment by PreziosiRaffaele** [comment]
It would be amazing to combine the user-friendly DevOps Center with hand-crafted pipelines!

**Comment by fmannreply** [comment]
I am unsure if this would improve our current workflow. We just started using the DevOps center and decided only Admins will use the DevOps Center, Devs still use our classic SFDX pipelines.

In the Status Quo:
1. Dev creates their changes in e.g. feature/123
2. Start deploy pipeline from feature/123 (from their sandbox) to develop (Review Sandbox)
3. If deployment succeeds the pipeline merges feature/123, otherwise not

So with this feature the suggested approach would be:
1. Devs create their changes in workitem 123
2. Start deploy pipeline from github to deploy workitem 123
3. `sf deploy pipeline` is triggered by the pipeline which then does the deployment via DevOps instead of Githubs action
4. Whats with the merge? When does that happen? Or is the merge supposed to trigger sf deploy pipeline? This would be a pattern we avoid. We never merge into develop if the code didn't deploy. It's like deploying code to develop that doesn't compile.

The only really advantage I see is it's using DevOps Center as deployment worker, instead of costing GitHub Action deployment time.

There are other issues I would like the DevOps Center to focus on. Duplication of Effort is one of the bigger issue. I'd rather see some `sf create workitem`. Devs shouldn't need to touch the DevOps Center.

**Comment by kfidelak94** [comment]
@fmannreply Thanks for the feedback. A couple thoughts/responses:
- The expected flow here is:
1. Work item is created in DevOps Center
2. Changes are committed to WI feature branch 
3. PR is created 
4. PR is merged
5. Changes are deployed using new CLI deploy pipeline command.

We do plan to add a CLI create work item command in future. It's in next set prioritized commands we're planning for the DevOps Center CLI plugin. Ahead of that, it could technically be scripted now by using the standard SF Rest API as it's simply a data record.  

Your point about not wanting to merge before you know the deploy is successful is interesting.  The metadata that you're deploying could be different before/after the merge.  So if you deploy before the merge you're not necessarily deploying the same thing that actually ends up in the post-merged branch (which is then what will progress to the next stage).  In DevOps Center UI, the "Promote" command actually first does the merge in memory, then does the deploy from that, and if the deploy succeeds then it commits the merge to the repo.  If the deploy does not succeed then it does not commit the merge so you're left back in the same state you were in before you attempted the promote.  So this addresses the concern of merging the branch before we know the deploy will be successful.  To get this same behavior from the CLI, this would be an argument for introducing the full/orchestrated "promote" command to the CLI, to support this same type of operation from the CLI.  With the new deploy command we're only performing the "deploy" part of the promotion, and assuming the "merge" has happened already separately.

**Comment by fmannreply** [comment]
@kfidelak94 Thanks for the response! So the idea would be once the "PR is merged", you could create a Github action, that would trigger the deployment via CLI, which then would properly promote the work item.

Great to here that you're looking into creating work items via CLI.

Very good point! I did work on a project once that didn't do this right. They determined the delta deployment via sgd, but then deployed the feature branch directly (spoiler - big mess as you pointed out as well). We do merge develop within the Github action into the feature and then deploy, only if the deployment succeed the pipeline pushes the updated develop branch.

With that we already have a pretty good hybrid flow. Downside being, that non-git users (Admins or Customer) can not track or see any of the work in the DevOps Center.

Maybe once the DevOps CLI expands, we could something like this:

1. Work item and branch is created in action, lets say `sf workitem create`
2. Changes are committed to WI feature branch
3. PR is created
4. Once approved, I can starting a "full/orchestrated promote". It syncs the work item, tries to deploy, when deployment succeeded merges PR, and promotes Work Item in DevOps Center. `sf workitem deploy --work-item WI-000002 --testlevel RunLocalTests`
